File: 03-march/19-count-submatrices-with-equal-frequency-of-x-and-y/solution.py
class Solution:
    def numberOfSubmatrices(self, grid: list[list[str]]) -> int:
        cols = len(grid[0])
        x_col = [0] * cols
        y_col = [0] * cols
        ans = 0

        for row in grid:
            x_prefix = 0
            y_prefix = 0
            for c, ch in enumerate(row):
                if ch == "X":
                    x_col[c] += 1
                elif ch == "Y":
                    y_col[c] += 1

                x_prefix += x_col[c]
                y_prefix += y_col[c]

                if x_prefix > 0 and x_prefix == y_prefix:
                    ans += 1

        return ans

File: 03-march/19-count-submatrices-with-equal-frequency-of-x-and-y/solution.java
class Solution {
    public int numberOfSubmatrices(char[][] grid) {
        int cols = grid[0].length;
        int[] xCol = new int[cols];
        int[] yCol = new int[cols];
        int ans = 0;

        for (char[] row : grid) {
            int xPrefix = 0;
            int yPrefix = 0;

            for (int c = 0; c < cols; c++) {
                if (row[c] == 'X') {
                    xCol[c]++;
                } else if (row[c] == 'Y') {
                    yCol[c]++;
                }

                xPrefix += xCol[c];
                yPrefix += yCol[c];

                if (xPrefix > 0 && xPrefix == yPrefix) {
                    ans++;
                }
            }
        }

        return ans;
    }
}
